### PR TITLE
session: Upgrade SDK package

### PR DESCRIPTION
**Pull request by cthulhu-rider**
Signed-off-by: Leonard Lyubich <leonard@nspcc.ru>

* based on and blocked by https://github.com/nspcc-dev/neofs-sdk-go/pull/248

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -96,3 +96,5 @@ require (
 	gopkg.in/yaml.v2 v2.4.0 // indirect
 	gopkg.in/yaml.v3 v3.0.0-20210107192922-496545a6307b // indirect
 )
+
+replace github.com/nspcc-dev/neofs-sdk-go => github.com/cthulhu-rider/neofs-sdk-go v0.0.0-20220523161616-ce50a783fcbf
```
Comment (fyrchik) [inline go.mod:100]: ?

File: pkg/services/object/acl/eacl/v2/xheader.go [hunk 2]
```diff
@@ -21,16 +21,21 @@ type responseXHeaderSource struct {
...
 
 	res := make([]eaclSDK.Header, 0, ln)
-	for meta := s.req.GetMetaHeader(); meta != nil; meta = meta.GetOrigin() {
-		x := meta.GetXHeaders()
-		for i := range x {
-			res = append(res, sessionSDK.NewXHeaderFromV2(&x[i]))
+
+	for i := range xHdrs {
+		for j := range xHdrs[i] {
+			res = append(res, (xHeader)(xHdrs[i][j]))
```
Comment (fyrchik) [inline pkg/services/object/acl/eacl/v2/xheader.go:38]: Why don't just replace a conversion? `sessionSDK.NewHeaderFromV2 -> x[i]`

File: pkg/services/object/acl/v2/util.go [hunk 3]
```diff
@@ -216,17 +216,30 @@ func isOwnerFromKey(id *user.ID, key *keys.PublicKey) bool {
...
+			sessionSDK.VerbObjectGet,
+			sessionSDK.VerbObjectDelete,
+			sessionSDK.VerbObjectRange,
+			sessionSDK.VerbObjectRangeHash)
+	case eaclSDK.OperationSearch:
+		return tok.AssertVerb(sessionSDK.VerbObjectSearch, sessionSDK.VerbObjectDelete)
+	case eaclSDK.OperationRange:
+		return tok.AssertVerb(sessionSDK.VerbObjectRange, sessionSDK.VerbObjectRangeHash)
+	case eaclSDK.OperationRangeHash:
+		return tok.AssertVerb(sessionSDK.VerbObjectRangeHash)
```
Comment (fyrchik) [inline pkg/services/object/acl/v2/util.go:241]: Why did you change the semantics?

File: pkg/services/object/acl/v2/util_test.go [hunk 4]
```diff
@@ -43,38 +48,48 @@ func testGenerateMetaHeader(depth uint32, b *acl.BearerToken, s *session.Token)
...
+		eacl.OperationHead: {
+			sessionSDK.VerbObjectHead,
+			sessionSDK.VerbObjectGet,
+			sessionSDK.VerbObjectDelete,
+			sessionSDK.VerbObjectRange,
+			sessionSDK.VerbObjectRangeHash,
+		},
+		eacl.OperationRange:     {sessionSDK.VerbObjectRange, sessionSDK.VerbObjectRangeHash},
+		eacl.OperationRangeHash: {sessionSDK.VerbObjectRangeHash},
+		eacl.OperationSearch:    {sessionSDK.VerbObjectSearch, sessionSDK.VerbObjectDelete},
```
Comment (fyrchik) [inline pkg/services/object/acl/v2/util_test.go:64]: It doesn't match the spec now.

File: go.mod (hunk 1 above)
Comment (cthulhu-rider) [reply go.mod:100]: See desc

File: pkg/services/object/acl/eacl/v2/xheader.go (hunk 2 above)
Comment (cthulhu-rider) [reply pkg/services/object/acl/eacl/v2/xheader.go:38]: Replace with what? We need `eaclSDK.Header` interface, `xHeader` provides it.

File: pkg/services/object/acl/v2/util_test.go (hunk 4 above)
Comment (cthulhu-rider) [reply pkg/services/object/acl/v2/util_test.go:64]: `Search` op can be executed within `Search` and `Delete` session, what mismatches?

File: pkg/services/object/util/prm.go [hunk 5]
```diff
@@ -102,17 +115,38 @@ func CommonPrmFromV2(req interface {
...
+
+	if tokenSessionV2 := meta.GetSessionToken(); tokenSessionV2 != nil {
+		ownerSessionV2 := tokenSessionV2.GetBody().GetOwnerID()
+		if ownerSessionV2 == nil {
+			return nil, errors.New("missing session owner")
+		}
+
+		err = ownerSession.ReadFromV2(*ownerSessionV2)
+		if err != nil {
+			return nil, fmt.Errorf("invalid session token: %w", err)
```
[the remark this answers is not in the document]
Comment (cthulhu-rider) [reply pkg/services/object/util/prm.go:130]: depends on https://github.com/nspcc-dev/neofs-sdk-go/pull/248 decisions.

**Review by carpawell — APPROVED** [review]
Except for @fyrchik's comments.

File: pkg/services/object/acl/eacl/v2/xheader.go (hunk 2 above)
Comment (fyrchik) [reply pkg/services/object/acl/eacl/v2/xheader.go:38]: I mean `xHdrs` slice was not here before, this is probably a rebase artifact.

File: pkg/services/object/acl/v2/util.go (hunk 3 above)
Comment (fyrchik) [reply pkg/services/object/acl/v2/util.go:241]: Oh, sorry, you now switch by op, and previously the switch was done by the token verb.

File: pkg/services/object/acl/v2/util_test.go (hunk 4 above)
Comment (fyrchik) [reply pkg/services/object/acl/v2/util_test.go:64]: Ok, I see, you swapped the order

File: pkg/services/object/acl/eacl/v2/xheader.go (hunk 2 above)
Comment (cthulhu-rider) [reply pkg/services/object/acl/eacl/v2/xheader.go]: Oh, got it, fixed.

**Review by carpawell — APPROVED** [review]
